## Thực hành 2: Làm quen với Python và SQL

### Python - Numpy

### NumPy là gì? Tại sao phải dùng NumPy?
NumPy (viết tắt của Numerical Python) là thư viện nền tảng cho mọi tính toán toán học và khoa học dữ liệu trong Python. Bản thân thư viện Pandas (giúp bạn tạo ra các DataFrame chứa bảng dữ liệu ở trên) cũng được xây dựng hoàn toàn trên nền tảng của NumPy.

* Tại sao không dùng List của Python? List của Python rất linh hoạt nhưng chạy cực kỳ chậm khi dữ liệu lên tới hàng triệu dòng. NumPy sử dụng ndarray (mảng N-chiều) được viết bằng ngôn ngữ C, giúp tính toán nhanh hơn gấp 100 đến 1000 lần và tiết kiệm bộ nhớ máy tính tối đa.

BÀI LAB 1: PHÂN TÍCH DỮ LIỆU BẢO TRÌ DỰ ĐOÁN BẰNG THƯ VIỆN NUMPY
Khởi tạo cấu hình và nạp dữ liệu ban đầuTrước khi đi vào các tác vụ phân tích, chúng ta cần khai báo đường dẫn tập dữ liệu và nạp dữ liệu từ file CSV vào bộ nhớ bằng thư viện Pandas, sau đó sẽ dùng NumPy để xử lý chuyên sâu.


In [ ]:
import numpy as np
import pandas as pd

In [ ]:
import pandas as pd
import numpy as np

# Đường dẫn đến các file dữ liệu (Cần chỉnh sửa lại cho đúng với máy của bạn)
file_ai4i = r"D:\NguyenMinhHuong\TA\SS_Bigdata\data\ai4i2020.csv"

# Đọc file dữ liệu AI4I 2020 Predictive Maintenance bằng Pandas
df_ai4i = pd.read_csv(file_ai4i)

Task 1: Chuyển đổi dữ liệu sang Mảng NumPy (np.array)
- Trong phân tích dữ liệu lớn và khoa học dữ liệu, để thực hiện các phép toán số học phức tạp với tốc độ cực nhanh, chúng ta cần chuyển đổi các cột dữ liệu đơn lẻ thành cấu trúc mảng của NumPy (được gọi là ndarray)
- Chức năng của hàm np.array(): Chuyển đổi một cấu trúc dữ liệu có sẵn (như List của Python hoặc Series của Pandas) thành một mảng NumPy (np.array).Cơ chế hoạt động, NumPy sẽ cấp phát một vùng nhớ liên tục để lưu trữ các phần tử có cùng kiểu dữ liệu. Điều này giúp tối ưu hóa bộ nhớ và tăng tốc độ tính toán lên gấp hàng trăm lần so với vòng lặp Python truyền thống.
- Cách dùng: np.array(du_lieu_dau_vao)

In [ ]:
# Lấy cột Nhiệt độ không khí và chuyển sang mảng NumPy
air_temp_array = np.array(df_ai4i['Air temperature [K]'])

print("--- TASK 1: CHUYỂN ĐỔI SANG MẢNG NUMPY ---")
print(f"- Đã chuyển cột 'Air temperature [K]' thành mảng NumPy.")
print(f"- Kiểu dữ liệu của mảng: {type(air_temp_array)}")
print(f"- 5 phần tử đầu tiên của mảng: {air_temp_array[:5]}\n")

--- TASK 1: CHUYỂN ĐỔI SANG MẢNG NUMPY ---
- Đã chuyển cột 'Air temperature [K]' thành mảng NumPy.
- Kiểu dữ liệu của mảng: <class 'numpy.ndarray'>
- 5 phần tử đầu tiên của mảng: [298.1 298.2 298.1 298.2 298.2]



Task 2: Phép toán Vector hóa (Tính toán đồng loạt không dùng vòng lặp)
- Bộ dữ liệu ghi nhận nhiệt độ của nhà máy theo thang đo độ Kelvin (°K). Tuy nhiên, tại Việt Nam và nhiều nước khác, các kỹ sư vận hành máy móc quen thuộc hơn với thang đo độ Celsius (°C). Chúng ta cần quy đổi toàn bộ hàng nghìn dòng dữ liệu nhiệt độ sang độ C.
- Chức năng của Vectorization: Thực hiện các phép toán số học (cộng, trừ, nhân, chia) trực tiếp trên toàn bộ mảng cùng một lúc mà không cần viết vòng lặp for. Cơ chế hoạt động, khi bạn lấy một mảng NumPy trừ đi một hằng số, NumPy tự động áp dụng phép trừ đó cho từng phần tử trong mảng (Cơ chế này gọi là Broadcasting).
- Cách dùng: mang_new = mang_cu - hang_so hoặc mang_moi = mang_1 + mang_2

In [ ]:
# Phép toán Vector hóa (Tính toán đồng loạt không dùng vòng lặp)
# Yêu cầu: Đổi toàn bộ mảng nhiệt độ từ độ K sang độ C (Công thức: Độ C = Độ K - 273.15)
air_temp_celsius = air_temp_array - 273.15
print(f"- 5 phần tử đầu tiên sau khi đổi sang độ C: {air_temp_celsius[:5]}\n")

- 5 phần tử đầu tiên sau khi đổi sang độ C: [24.95 25.05 24.95 25.05 25.05]



Task 3: Tính toán Thống kê Cơ bản với np.mean()
- Để đánh giá môi trường làm việc tổng thể của nhà máy có quá nóng hay không, người quản lý cần biết giá trị nhiệt độ trung bình là bao nhiêu. Điều này giúp đưa ra quyết định có nên lắp đặt thêm hệ thống làm mát trung tâm hay không.
- Chức năng của hàm np.mean(): Tính giá trị trung bình cộng (Mean/Average) của tất cả các phần tử số có trong mảng dữ liệu.Cơ chế hoạt động, hàm sẽ cộng tổng tất cả các phần tử trong mảng lại, sau đó chia cho tổng số lượng phần tử.
- Cách dùng: np.mean(mang_du_lieu)

In [ ]:
#Thống kê cơ bản với np.mean() và np.sum()
# Tính nhiệt độ trung bình của nhà máy
mean_temp = np.mean(air_temp_array)
print(f"- Nhiệt độ không khí trung bình (độ K): {mean_temp:.2f} K")
print(f"- Nhiệt độ không khí trung bình (độ C): {np.mean(air_temp_celsius):.2f} °C")

- Nhiệt độ không khí trung bình (độ K): 300.00 K
- Nhiệt độ không khí trung bình (độ C): 26.85 °C


Task 4: Tìm giá trị Cực trị bằng np.max() và np.min()
- Mỗi loại máy móc trong nhà máy đều có một giới hạn vận hành an toàn về tốc độ quay (Rotational speed). Nếu máy quay quá chậm, năng suất sẽ giảm. Ngược lại, nếu máy quay quá nhanh vượt ngưỡng thiết kế, linh kiện sẽ bị mài mòn rất nhanh và có nguy cơ nổ vỡ. Chúng ta cần tìm ra các giá trị biên (lớn nhất và nhỏ nhất) đã từng xuất hiện trong lịch sử vận hành.
- Chức năng của hàm np.max() & np.min(): Tìm ra giá trị lớn nhất (np.max) và giá trị nhỏ nhất (np.min) có trong mảng dữ liệu.
- Cách dùng:
    + np.max(mang_du_lieu): Trả về phần tử có giá trị cao nhất
    + np.min(mang_du_lieu): Trả về phần tử có giá trị thấp nhất.

In [ ]:
# Tìm giá trị Cực trị bằng np.max() và np.min()
rot_speed_array = np.array(df_ai4i['Rotational speed [rpm]'])

max_speed = np.max(rot_speed_array)
min_speed = np.min(rot_speed_array)

print(f"- Tốc độ quay lớn nhất của máy móc (np.max): {max_speed} rpm")
print(f"- Tốc độ quay nhỏ nhất của máy móc (np.min): {min_speed} rpm\n")

- Tốc độ quay lớn nhất của máy móc (np.max): 2886 rpm
- Tốc độ quay nhỏ nhất của máy móc (np.min): 1168 rpm



Task 5: Tính Độ lệch chuẩn bằng np.std() để đo độ ổn định
- Lực xoắn (Torque) tác động lên trục máy là một chỉ số cần sự ổn định. Nếu lực xoắn lúc quá cao, lúc quá thấp (trồi sụt thất thường), máy sẽ bị rung lắc mạnh và dễ hỏng. Chỉ biết giá trị trung bình là chưa đủ, kỹ sư cần biết dữ liệu có bị phân tán, dao động mạnh hay không.
- Chức năng của hàm np.std(): Tính độ lệch chuẩn (Standard Deviation) của mảng dữ liệu. Đây là thước đo mức độ phân tán của các giá trị xung quanh số trung bình.Ý nghĩa thực tế:Độ lệch chuẩn nhỏ: Các giá trị rất sát với mức trung bình → Máy hoạt động ổn định.Độ lệch chuẩn lớn: Các giá trị dao động xa mức trung bình → Máy hoạt động chập chờn, bất ổn định.
- Cách dùng: np.std(mang_du_lieu)

In [ ]:
# Tính Độ lệch chuẩn bằng np.std() để đo độ ổn định
torque_array = np.array(df_ai4i['Torque [Nm]'])

std_torque = np.std(torque_array)
print(f"- Lực xoắn trung bình: {np.mean(torque_array):.2f} Nm")
print(f"- Độ phân tán/Độ lệch chuẩn lực xoắn (np.std): {std_torque:.2f} Nm")
print("- Nhận xét: Lực xoắn của máy thường biến thiên xung quanh mức trung bình +/- độ lệch chuẩn này.\n")


- Lực xoắn trung bình: 39.99 Nm
- Độ phân tán/Độ lệch chuẩn lực xoắn (np.std): 9.97 Nm
- Nhận xét: Lực xoắn của máy thường biến thiên xung quanh mức trung bình +/- độ lệch chuẩn này.



Task 6: Đếm số lượng sự cố bằng np.sum() kết hợp Điều kiện
- Cột dữ liệu Machine failure (Hỏng máy) chứa các giá trị nhị phân: 1 đại diện cho việc máy bị hỏng tại thời điểm đó, và 0 là máy hoạt động bình thường. Chúng ta cần đếm tổng số ca hỏng máy để báo cáo cho phòng bảo trì. Đồng thời, cần điều tra sâu hơn xem có bao nhiêu ca hỏng máy xảy ra khi thời tiết quá oi bức (Nhiệt độ môi trường > 300°K).
- Chức năng của np.sum() kết hợp Mặt nạ Điều kiện (Boolean Mask): Tính tổng các phần tử trong mảng. Khi kết hợp với biểu thức logic (như >, <, ==), NumPy sẽ tạo ra mảng chứa các giá trị True (đúng) và False (sai).
- Cơ chế hoạt động: Trong toán học máy tính, True tương đương với 1 và False tương đương với 0. Vì vậy, khi dùng np.sum() trên một mảng điều kiện, bản chất là chúng ta đang cộng tất cả các giá trị 1 (True) lại, từ đó ra được số lượng phần tử thỏa mãn điều kiện.
- Phép toán logic nhiều điều kiện: Sử dụng dấu & cho điều kiện ĐỒNG THỜI XẢY RA (VÀ) và dấu | cho điều kiện HOẶC. Lưu ý phải bọc các mệnh đề trong dấu ngoặc đơn ()

In [ ]:
# Đếm số lượng sự cố bằng np.sum() kết hợp Điều kiện
failure_array = np.array(df_ai4i['Machine failure'])

# Tính tổng số ca hỏng máy (Cột failure nhận giá trị 0 hoặc 1)
total_failures = np.sum(failure_array)
print(f"- Tổng số ca thiết bị gặp sự cố hỏng hóc trong nhà máy: {total_failures} ca")

# Đếm xem có bao nhiêu máy đang hoạt động ở mức nhiệt độ quá cao (> 300 K) mà bị hỏng
high_temp_failures = np.sum((air_temp_array > 300) & (failure_array == 1))
print(f"- Số ca hỏng máy xảy ra khi nhiệt độ môi trường vượt quá 300K: {high_temp_failures} ca\n")

- Tổng số ca thiết bị gặp sự cố hỏng hóc trong nhà máy: 339 ca
- Số ca hỏng máy xảy ra khi nhiệt độ môi trường vượt quá 300K: 230 ca



Task 7: Tính Tỷ lệ phần trăm sự cố bằng np.mean()
- Thay vì chỉ biết số ca hỏng thô (ví dụ: 339 ca), ban giám đốc cần một con số mang tính so sánh chung như: "Tỷ lệ hỏng hóc của nhà máy chiếm bao nhiêu phần trăm trên tổng số số giờ vận hành?". Con số phần trăm này giúp doanh nghiệp so sánh mức độ an toàn với các nhà máy khác trong ngành.
- Chức năng và Cách dùng mẹo với hàm np.mean(): Đối với một mảng dữ liệu nhị phân chỉ toàn số 0 và 1, hàm np.mean() ngoài tính trung bình cộng thì nó phản ánh chính xác tỷ lệ xuất hiện của các số 1.Cơ chế hoạt động: Trung bình = (Tổng các số 1) / (Tổng số lượng phần tử 0 và 1). Đây chính là công thức tính tỷ lệ phần trăm phần tử đạt điều kiện.
- Cách dùng: np.mean(mang_nhi_phan) rồi nhân với 100 để đổi ra đơn vị %.

In [ ]:
# Tính Tỷ lệ phần trăm sự cố bằng np.mean()
# Với mảng chứa giá trị nhị phân 0 và 1, hàm np.mean() chính là tỷ lệ xuất hiện của giá trị 1
failure_rate = np.mean(failure_array)

print(f"- Tỷ lệ hỏng hóc trung bình của toàn bộ nhà máy: {failure_rate:.4f}")
print(f"- Quy đổi sang phần trăm: {failure_rate * 100:.2f}%\n")

- Tỷ lệ hỏng hóc trung bình của toàn bộ nhà máy: 0.0339
- Quy đổi sang phần trăm: 3.39%



Task 8: Lọc dữ liệu mảng bằng kỹ thuật Boolean Indexing
- Khi hệ thống cảnh báo có hàng trăm máy gặp sự cố, kỹ sư không thể ngồi dò từng dòng để xem cấu hình vận hành của riêng những máy hỏng đó. Ban kỹ thuật cần tách ngay toàn bộ dữ liệu tốc độ quay của nhóm máy lỗi ra một danh sách riêng. Mục tiêu là phân tích xem có phải tốc độ quá cao là nguyên nhân gây hỏng hóc hay không.

- Chức năng đặc biệt: Kỹ thuật dùng một mảng điều kiện Đúng/Sai (True/False) làm "bộ lọc" để trích xuất các phần tử từ một mảng khác có cùng kích thước.
* Cơ chế hoạt động: Khi bạn truyền biểu thức failure_array == 1 vào trong dấu ngoặc vuông của mảng tốc độ, NumPy sẽ quét qua từng vị trí. Chỉ những vị trí nào ứng với số 1 (Máy hỏng - True) thì giá trị tốc độ tại đó mới được giữ lại.
* Cách dùng: mang_can_loc[mang_dieu_kien == gia_tri]. Kết quả trả về là một mảng mới ngắn hơn, chỉ chứa dữ liệu thỏa mãn điều kiện.


In [ ]:
# Lọc dữ liệu mảng bằng kỹ thuật Boolean Indexing
# Lọc lấy danh sách tốc độ quay của riêng những máy đã bị hỏng
speed_of_failed_machines = rot_speed_array[failure_array == 1]

print(f"- Số lượng mẫu máy hỏng thu được: {len(speed_of_failed_machines)}")
print(f"- Tốc độ quay trung bình của riêng nhóm máy bị hỏng này: {np.mean(speed_of_failed_machines):.2f} rpm\n")

- Số lượng mẫu máy hỏng thu được: 339
- Tốc độ quay trung bình của riêng nhóm máy bị hỏng này: 1496.49 rpm



Task 9: Phân nhóm dữ liệu bằng np.percentile() để tìm ngoại lai
- Để thiết lập ngưỡng cảnh báo sớm cho thiết bị, phòng bảo trì không thể chọn đại một con số cảm tính. Họ cần biết mức lực xoắn (torque) thực tế mà đại đa số các máy đang chịu đựng là bao nhiêu. Việc xác định được mốc giới hạn mà 95% số máy vận hành dưới mức đó giúp doanh nghiệp cô lập được 5% nhóm máy "cá biệt" đang chạy quá tải để lên lịch bảo dưỡng nghiêm ngặt.

- Chức năng và Cách dùng mẹo với hàm np.percentile()
* Chức năng đặc biệt: Hàm giúp tìm giá trị phân vị (Percentile), tức là điểm mốc chia dữ liệu thành hai phần theo tỷ lệ phần trăm mong muốn sau khi đã sắp xếp từ thấp đến cao.
* Cơ chế hoạt động: Khi truyền vào con số 95, NumPy sẽ tính toán ra một giá trị lực xoắn cụ thể. Tại giá trị này, dữ liệu sẽ chia làm 2 phe rõ rệt: 95% số máy có lực xoắn thấp hơn hoặc bằng nó, và chỉ 5% số máy có lực xoắn cao hơn nó.
* Cách dùng: np.percentile(mang_du_lieu, phan_tram). Trong đó phan_tram nhận giá trị từ 0 đến 100.

In [ ]:
# Phân nhóm dữ liệu bằng np.percentile() để tìm ngoại lai
# Tìm mức giới hạn mà 95% số máy có lực xoắn thấp hơn mức này
torque_95th = np.percentile(torque_array, 95)

print(f"- Mức phân vị thứ 95 của lực xoắn: {torque_95th:.2f} Nm")
print(f"- Ý nghĩa: Chỉ có 5% số máy trong nhà máy vận hành với lực xoắn cực đại vượt quá {torque_95th:.2f} Nm (Nguy cơ quá tải).\n")


- Mức phân vị thứ 95 của lực xoắn: 56.10 Nm
- Ý nghĩa: Chỉ có 5% số máy trong nhà máy vận hành với lực xoắn cực đại vượt quá 56.10 Nm (Nguy cơ quá tải).



Task 10: Làm sạch dữ liệu bằng np.clip() (Giới hạn biên)
- Cảm biến nhiệt độ hoặc tốc độ trong nhà máy đôi khi bị nhiễu do sóng từ trường, sinh ra các giá trị ảo cực đoan (ví dụ tốc độ vọt lên 5000 rpm hoặc âm vài trăm rpm trong tích tắc). Nếu đưa trực tiếp dữ liệu rác này vào mô hình AI, máy tính sẽ dự báo sai hoàn toàn. Doanh nghiệp cần một "hàng rào kỹ thuật" để ép các giá trị vượt khung này về đúng biên độ thực tế quy định (1200 rpm đến 2000 rpm).

- Chức năng và Cách dùng mẹo với hàm np.clip()
* Chức năng đặc biệt: Hàm này hoạt động như một cái khuôn cắt biên dữ liệu, tự động "gọt" phẳng các phần tử vượt ngưỡng mà không làm thay đổi các giá trị đã nằm trong vùng an toàn.
* Cơ chế hoạt động: NumPy duyệt qua mảng: phần tử nào nhỏ hơn ngưỡng dưới sẽ bị ép bằng ngưỡng dưới; phần tử nào lớn hơn ngưỡng trên sẽ bị ép bằng ngưỡng trên; các phần tử ở giữa được giữ nguyên.
* Cách dùng: np.clip(mang_du_lieu, nguong_duoi, nguong_tren). Đây là giải pháp làm sạch dữ liệu cực nhanh thay vì phải viết nhiều vòng lặp if-else phức tạp.

In [ ]:
# Làm sạch dữ liệu bằng np.clip() (Giới hạn biên)
# Giả sử quy định tốc độ quay không được quá 2000 rpm và không dưới 1200 rpm để tránh nhiễu cảm biến
cleaned_speed = np.clip(rot_speed_array, 1200, 2000)

print(f"- Tốc độ tối đa ban đầu: {np.max(rot_speed_array)} rpm")
print(f"- Tốc độ tối đa sau khi xử lý cắt biên (np.clip): {np.max(cleaned_speed)} rpm")

- Tốc độ tối đa ban đầu: 2886 rpm
- Tốc độ tối đa sau khi xử lý cắt biên (np.clip): 2000 rpm


### BÀI TẬP NUMPY: PHÂN TÍCH THÔNG SỐ CẢM BIẾN MÁY MÓC
Giả sử ta có dữ liệu đo đạc của 8 máy trong xưởng sản xuất:
* Nhiệt độ không khí (°K): `temp_list = [298.1, 298.5, 301.2, 299.0, 302.5, 297.8, 303.1, 298.0]`
* Thời gian mài mòn dao (phút): `tool_wear_list = [10, 45, 180, 210, 5, 240, 195, 30]`
* Trạng thái hỏng máy (0: Bình thường, 1: Hỏng): `fail_list = [0, 0, 1, 1, 0, 1, 1, 0]`


#### Yêu cầu:
1. *Bài 1 (Khởi tạo & Vector hóa)*: Chuyển `temp_list` thành mảng NumPy `temp_array`. Sau đó chuyển đổi toàn bộ mảng sang độ °C (Công thức: `°C = °K - 273.15`) mà *không dùng vòng lặp for*.
2. *Bài 2 (Thống kê cơ bản)*: Tính nhiệt độ trung bình (`np.mean`), cao nhất (`np.max`), thấp nhất (`np.min`) và độ lệch chuẩn (`np.std`) của mảng nhiệt độ °C vừa tính.
3. *Bài 3 (Đếm điều kiện)*: Đếm xem có bao nhiêu máy có nhiệt độ lớn hơn hoặc bằng `300°K` (Gợi ý: dùng `np.sum(mảng_điều_kiện)`).
4. *Bài 4 (Lọc Boolean Indexing)*: Lọc ra danh sách thời gian mòn dao (`tool_wear`) của *riêng những máy bị hỏng* (`fail_list == 1`).
5. *Bài 5 (Cắt biên np.clip)*: Thời gian mài mòn dao chuẩn không được vượt quá `200` phút. Hãy dùng `np.clip()` để ép các giá trị lớn hơn 200 về mức trần `200`.


In [ ]:
import numpy as np

# Dữ liệu ban đầu
temp_list = [298.1, 298.5, 301.2, 299.0, 302.5, 297.8, 303.1, 298.0]
tool_wear_list = [10, 45, 180, 210, 5, 240, 195, 30]
fail_list = [0, 0, 1, 1, 0, 1, 1, 0]

# --- BÀI 1: Khởi tạo mảng & Đổi sang độ C ---
temp_array =
temp_celsius =
print("Nhiệt độ độ C:", temp_celsius)

# --- BÀI 2: Thống kê cơ bản ---
print(f"Nhiệt độ TB: {np.mean(...):.2f} °C")
print(f"Nhiệt độ Max: {np.max(...):.2f} °C, Min: {np.min(...):.2f} °C")
print(f"Độ lệch chuẩn: {np.std(...):.2f}")

# --- BÀI 3: Đếm số máy có nhiệt độ >= 300 K ---
count_high_temp =
print(f"Số máy có nhiệt độ >= 300K: {count_high_temp}")

# --- BÀI 4: Lọc thời gian mòn dao của các máy bị hỏng (fail == 1) ---
tool_wear_array = np.array(tool_wear_list)
fail_array = np.array(fail_list)

wear_failed_machines =
print("Thời gian mòn dao của các máy hỏng:", wear_failed_machines)

# --- BÀI 5: Giới hạn biên với np.clip (Tối đa 200 phút) ---
clipped_tool_wear =
print("Thời gian mòn dao sau khi cắt biên:", clipped_tool_wear)


SyntaxError: invalid syntax (2547930313.py, line 9)

### Python - Pandas

### Pandas là gì? Tại sao phải dùng Pandas?
Pandas (viết tắt của Panel Data) là thư viện tối quan trọng trong Python chuyên dùng để thao tác, làm sạch và phân tích dữ liệu dưới dạng bảng. Nếu NumPy cung cấp sức mạnh tính toán mảng mạnh mẽ ở bên dưới, thì Pandas chính là giao diện "Excel chạy bằng code" giúp bạn dễ dàng quản lý hàng triệu dòng dữ liệu có cấu trúc.

* Tại sao không dùng Excel hay SQL? Excel sẽ bị treo hoặc sập nguồn khi tệp dữ liệu vượt quá 1 triệu dòng và không thể tự động hóa các tác vụ phức tạp. SQL rất mạnh để truy vấn nhưng lại hạn chế trong việc làm sạch dữ liệu chuyên sâu và trực quan hóa đồ thị. Pandas kết hợp sức mạnh của cả hai: xử lý dữ liệu khổng lồ với tốc độ cực nhanh, quản lý linh hoạt bằng hai cấu trúc chính là Series (mảng 1 chiều có nhãn) và DataFrame (bảng 2 chiều có hàng và cột).

BÀI LAB 2: THAO TÁC VÀ LÀM SẠCH DỮ LIỆU NHÀ MÁY BẰNG THƯ VIỆN PANDAS
- Khởi tạo cấu hình và nạp dữ liệu ban đầu: Để bắt đầu chiến dịch chuẩn hóa dữ liệu vận hành, chúng ta sẽ sử dụng Pandas để đọc trực tiếp file dữ liệu thô (CSV/Excel). Thư viện này sẽ tự động chuyển đổi các hàng và cột thô thành một DataFrame trực quan, sẵn sàng cho các bước lọc nhiễu, xử lý dữ liệu trống (Missing data) và trích xuất thông tin nghiệp vụ.


In [ ]:
import os
import pandas as pd

# Đọc và in ra 5 hàng đầu tiên (df.head())
df = pd.read_csv(file_ai4i)

# In ra 5 hàng đầu tiên
df.head()


,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


Task 1: Đọc và kiểm tra nhanh dữ liệu bằng pd.read_csv() và df.head()
- Trước khi bắt tay vào phân tích hay xây dựng mô hình AI, bước đầu tiên và quan trọng nhất là phải "nhìn thấy" dữ liệu. Bạn cần biết tập dữ liệu gồm những cột nào, kiểu dữ liệu ra sao và hình thù các dòng đầu tiên như thế nào. Việc in ra một vài hàng đầu tiên giúp kỹ sư xác nhận dữ liệu đã được nạp đúng cấu trúc, không bị lỗi font hay lệch cột.
Chức năng và Cách dùng mẹo với hàm trong Pandas

* Hàm pd.read_csv():
* Chức năng: Tự động quét qua file CSV, nhận diện các hàng, cột và dấu phân cách để chuyển đổi toàn bộ tệp văn bản thô đó thành một bảng dữ liệu thông minh (DataFrame).
   * Cách dùng: pd.read_csv(duong_dan_file).
* Phương thức df.head():
* Chức năng: Trích xuất nhanh một số dòng đầu tiên của bảng dữ liệu để xem trước (mặc định là 5 dòng).
   * Cơ chế hoạt động: Thay vì bắt máy tính phải hiển thị toàn bộ hàng triệu dòng gây lag màn hình, df.head() chỉ cắt lấy phần ngọn của bảng, giúp bạn kiểm tra nhanh cấu trúc dữ liệu mà không tốn tài nguyên.
   * Cách dùng: df.head(so_dong) (Nếu để trống sẽ tự động lấy 5 dòng, hoặc bạn có thể điền số dòng mong muốn như df.head(10)).

In [ ]:
# Quan sát cấu trúc dữ liệu với df.shape và df.info()
# df.shape trả về (số dòng, số cột)
print(f"- Kích thước tập dữ liệu (df.shape): {df.shape[0]} dòng và {df.shape[1]} cột.\n")

# df.info() in ra kiểu dữ liệu của từng cột và số lượng dòng có dữ liệu (Non-Null)
print("- Chi tiết cấu trúc các cột (df.info()):")
df.info()
print("\n")

- Kích thước tập dữ liệu (df.shape): 10000 dòng và 14 cột.

- Chi tiết cấu trúc các cột (df.info()):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   UDI                      10000 non-null  int64  
 1   Product ID               10000 non-null  object 
 2   Type                     10000 non-null  object 
 3   Air temperature [K]      10000 non-null  float64
 4   Process temperature [K]  10000 non-null  float64
 5   Rotational speed [rpm]   10000 non-null  int64  
 6   Torque [Nm]              10000 non-null  float64
 7   Tool wear [min]          10000 non-null  int64  
 8   Machine failure          10000 non-null  int64  
 9   TWF                      10000 non-null  int64  
 10  HDF                      10000 non-null  int64  
 11  PWF                      10000 non-null  int64  
 12  OSF                      10000

In [ ]:
# Thống kê mô tả đơn giản bằng df.describe()
# Tự động tính toán cho tất cả các cột số thực/số nguyên trong bảng
df.describe()

,UDI,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000
mean,5000.50000,300.004930,310.005560,1538.776100,39.986910,107.951000,0.033900,0.004600,0.011500,0.009500,0.009800,0.00190
std,2886.89568,2.000259,1.483734,179.284096,9.968934,63.654147,0.180981,0.067671,0.106625,0.097009,0.098514,0.04355
min,1.00000,295.300000,305.700000,1168.000000,3.800000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
25%,2500.75000,298.300000,308.800000,1423.000000,33.200000,53.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
50%,5000.50000,300.100000,310.100000,1503.000000,40.100000,108.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
75%,7500.25000,301.500000,311.100000,1612.000000,46.800000,162.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000
max,10000.00000,304.500000,313.800000,2886.000000,76.600000,253.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000


In [ ]:
# Thống kê giá trị trung bình cụ thể bằng df['cột'].mean()
# Tính nhiệt độ không khí trung bình
mean_air_temp = df['Air temperature [K]'].mean()
print(f"- Nhiệt độ không khí trung bình (độ K): {mean_air_temp:.2f} K")

# Tạo cột mới đổi sang độ C trực tiếp trên DataFrame và tính trung bình
df['Air temperature [C]'] = df['Air temperature [K]'] - 273.15
print(f"- Nhiệt độ không khí trung bình (độ C): {df['Air temperature [C]'].mean():.2f} °C\n")


- Nhiệt độ không khí trung bình (độ K): 300.00 K
- Nhiệt độ không khí trung bình (độ C): 26.85 °C



In [ ]:
# Đếm tổng số ca hỏng máy bằng df['cột'].sum()
total_failures = df['Machine failure'].sum()
print(f"- Tổng số ca máy bị hỏng ghi nhận trong nhà máy (df.sum): {total_failures} ca\n")

- Tổng số ca máy bị hỏng ghi nhận trong nhà máy (df.sum): 339 ca



In [ ]:
# Thống kê tần suất danh mục bằng df['cột'].value_counts()
# Lấy ra ký tự đầu tiên (L, M, H) trong cột Product ID để đếm loại máy
df['Type'] = df['Product ID'].str[0]
type_counts = df['Type'].value_counts()

print("- Số lượng máy móc theo từng phân cấp chất lượng:")
for product_type, count in type_counts.items():
    print(f"  + Loại {product_type} (Low/Medium/High): {count} máy")
print("\n")


- Số lượng máy móc theo từng phân cấp chất lượng:
  + Loại L (Low/Medium/High): 6000 máy
  + Loại M (Low/Medium/High): 2997 máy
  + Loại H (Low/Medium/High): 1003 máy




In [ ]:
# Phân nhóm dữ liệu và tính trung bình bằng df.groupby()
# Gom nhóm theo trạng thái hỏng (0 và 1) rồi tính trung bình các cột cảm biến chính
grouped_summary = df.groupby('Machine failure')[['Air temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]']].mean()
grouped_summary
print("- Nhận xét: Máy bị hỏng thường có lực xoắn (Torque) và thời gian mòn dao (Tool wear) trung bình cao hơn máy bình thường.\n")


- Nhận xét: Máy bị hỏng thường có lực xoắn (Torque) và thời gian mòn dao (Tool wear) trung bình cao hơn máy bình thường.



In [ ]:
# Thống kê chi tiết 5 loại lỗi cụ thể (Dữ liệu lớn)
failure_types = ['TWF', 'HDF', 'PWF', 'OSF', 'RNF']
print("- Tổng số ca hỏng của từng loại lỗi cụ thể:")
for fault in failure_types:
    print(f"  + Lỗi {fault}: {df[fault].sum()} ca")

- Tổng số ca hỏng của từng loại lỗi cụ thể:
  + Lỗi TWF: 46 ca
  + Lỗi HDF: 115 ca
  + Lỗi PWF: 95 ca
  + Lỗi OSF: 98 ca
  + Lỗi RNF: 19 ca


### BÀI TẬP PANDAS: PHÂN TÍCH TẬP DỮ LIỆU CẢM BIẾN AI4I 2020
Sử dụng file dữ liệu `ai4i2020.csv` đã nạp vào biến `df` (hoặc `df_ai4i`) để thực hiện các yêu cầu sau:

---
#### Yêu cầu:
1. *Bài 1 (Đọc & Khám phá)*: Đọc file `ai4i2020.csv` vào biến `df`, in ra kích thước của bảng (`df.shape`) và hiển thị 5 dòng đầu tiên (`df.head()`).
2. *Bài 2 (Tạo cột mới & Tính trung bình)*: Tạo cột mới tên là `Process temperature [C]` bằng cách đổi nhiệt độ quy trình từ độ K sang độ C (`Process temperature [K] - 273.15`). Sau đó tính giá trị trung bình (`.mean()`) của cột mới này.
3. *Bài 3 (Đếm phân loại `value_counts`)*: Đếm số lượng máy móc theo từng phân cấp chất lượng trong cột `Type` (gồm các loại L, M, H).
4. *Bài 4 (Lọc điều kiện)*: Lọc ra danh sách các máy có tốc độ quay cực cao: `Rotational speed [rpm] > 2500` và in ra số lượng máy thỏa mãn điều kiện này.
5. *Bài 5 (Gom nhóm `groupby`)*: Gom nhóm theo cột `Type` và tính lực xoắn trung bình (`Torque [Nm]`) của từng loại máy.
